File: WinForms/VB/NavBarMainDemo/Modules/MultimediaModule.vb
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Drawing.Imaging
Imports System.IO
Imports DevExpress.DXperience.Demos
Imports DevExpress.Utils.Drawing
Imports DevExpress.XtraBars.Ribbon
Imports DevExpress.XtraEditors

Namespace DevExpress.XtraNavBar.Demos

    Public Partial Class AccordionControlMultimediaModule
        Inherits TutorialControl

        Public Sub New()
            InitializeComponent()
            WorkingFolder = DataPath
            If Not String.IsNullOrEmpty(WorkingFolder) Then
                InitializeContent()
                SetImage(Images(0))
            End If
        End Sub

        Protected Overridable Sub InitializeContent()
            InitializeTileControl()
            UpdateEditors()
            AddHandler galleryFilters.SizeChanged, AddressOf OnGalleryFiltersSizeChanged
        End Sub

        Protected Overridable Sub InitializeTileControl()
            TryCast(tileControl1, ITileControl).ViewInfo.UseAdvancedTextRendering = False
            If TileGroupImages Is Nothing Then Return
            TileControl.Groups.Clear()
            TileControl.Groups.Add(TileGroupImages)
        End Sub

        Protected Overridable Sub InitializeImages()
            If String.IsNullOrEmpty(WorkingFolder) Then Return
            Images = New List(Of Image)()
            Dim files As String() = Directory.GetFiles(WorkingFolder, "*.jpg")
            For Each file As String In files
                Dim img As Image = Bitmap.FromFile(file)
                img.Tag = file
                Images.Add(img)
            Next

            InitializeTileControl()
        End Sub

        Protected Overridable Property Images As List(Of Image)

        Protected Overridable ReadOnly Property TileControl As TileControl
            Get
                Return tileControl1
            End Get
        End Property

        Protected Overridable ReadOnly Property PictureEdit As PictureEdit
            Get
                Return peCurrentImage
            End Get
        End Property

        Protected Overridable ReadOnly Property FilterGroup As GalleryItemGroup
            Get
                Return galleryFilters.Gallery.Groups(0)
            End Get
        End Property

        Protected Overridable ReadOnly Property PreviewSize As Size
            Get
                Return galleryFilters.Gallery.ImageSize
            End Get
        End Property

        Private originalImageCore As Image

        Protected Overridable Property OriginalImage As Image
            Get
                Return originalImageCore
            End Get

            Set(ByVal value As Image)
                originalImageCore = value
                peOriginalImage.Image = originalImageCore
                UpdateImageProperties()
            End Set
        End Property

        Private currentImageCore As Image

        Protected Overridable Property CurrentImage As Image
            Get
                Return currentImageCore
            End Get

            Set(ByVal value As Image)
                currentImageCore = value
                If CurrentImage Is Nothing Then Return
                PictureEdit.Image = CurrentImage
                CalcCurrentImageZoomPercent(currentImageCore)
                UpdateEditors()
            End Set
        End Property

        Protected Overridable Sub CalcCurrentImageZoomPercent(ByVal img As Image)
            PictureEdit.Properties.ZoomPercent = Math.Min((CDbl(PictureEdit.Height) / CDbl(img.Height)) * 100, 100)
        End Sub

        Private workingFolderCore As String

        Protected Overridable Property WorkingFolder As String
            Get
                Return workingFolderCore
            End Get

            Set(ByVal value As String)
                If Equals(workingFolderCore, value) Then Return
                workingFolderCore = value
                InitializeImages()
            End Set
        End Property

        Private group As TileGroup

        Protected Overridable ReadOnly Property TileGroupImages As TileGroup
            Get
                If group Is Nothing Then group = CreateTileGroupImages()
                Return group
            End Get
        End Property

        Protected Overridable Sub SetImage(ByVal img As Image)
            OriginalImage = img
            CurrentImage = OriginalImage
            RefreshFilters(OriginalImage)
        End Sub

        Protected ReadOnly Property DataPath As String
            Get
                Dim dataDir As String = Tutorials.FilePathUtils.FindDirPath("Data")
                Dim dataPathCore As String = Path.Combine(dataDir, "AccordionControlData")
                If Directory.Exists(dataPathCore) Then Return dataPathCore
                Return String.Empty
            End Get
        End Property

        Protected Overridable Sub UpdateEditors()
            Dim value As Boolean = CurrentImage IsNot Nothing
            tbRed.Enabled = value
            tbGreen.Enabled = value
            tbBlue.Enabled = value
            tbBrightness.Enabled = value
            tbContrast.Enabled = value
            tbRed.Value = tbRed.Properties.Minimum
            tbGreen.Value = tbGreen.Properties.Minimum
            tbBlue.Value = tbBlue.Properties.Minimum
            tbBrightness.Value = tbBrightness.Properties.Minimum
            tbContrast.Value = tbContrast.Properties.Minimum
        End Sub

        Protected Overridable Function CreateTileGroupImages() As TileGroup
            If Images Is Nothing OrElse Images.Count = 0 Then Return Nothing
            Dim g As TileGroup = New TileGroup()
            For Each img As Image In Images
                Dim item As TileItem = New TileItem() With {.BackgroundImage = img, .BackgroundImageScaleMode = TileItemImageScaleMode.Squeeze, .BackgroundImageAlignment = TileItemContentAlignment.MiddleCenter}
                g.Items.Add(item)
            Next

            Return g
        End Function

        Protected Overridable Sub RefreshFilters(ByVal img As Image)
            FilterGroup.Items.Clear()
            For Each filter As FilterBase In FilterList
                FilterGroup.Items.Add(CreateGalleryItemFilter(img, filter))
            Next
        End Sub

        Private updatingAccordionWidth As Boolean

        Protected Overridable Sub OnGalleryFiltersSizeChanged(ByVal sender As Object, ByVal e As EventArgs)
            UpdateAccordionPanelWidth()
        End Sub

        Protected Overridable Sub UpdateAccordionPanelWidth()
            If updatingAccordionWidth Then Return
            If FilterGroup.Items.Count = 0 Then Return
            updatingAccordionWidth = True
            Try
                Dim bestSize As Size = galleryFilters.GetPreferredSize(Size.Empty)
                Dim delta As Integer = bestSize.Width - galleryFilters.Width
                If delta = 0 Then Return
                panelControl7.Width += delta
            Finally
                updatingAccordionWidth = False
            End Try
        End Sub

        Protected Overridable Sub UpdateImageProperties()
            ratingImage.Rating = GetRating(OriginalImage)
            lblSize.Text = GetSize(OriginalImage)
            lblName.Text = GetName(OriginalImage)
            lblDimension.Text = GetDimension(OriginalImage)
        End Sub

        Protected Overridable Function CreateGalleryItemFilter(ByVal img As Image, ByVal filter As FilterBase) As GalleryItem
            Dim item As GalleryItem = New GalleryItem() With {.Image = filter.ApplyFilter(img, PreviewSize), .Caption = filter.Name}
            AddHandler item.ItemClick, Sub() PictureEdit.Image = filter.ApplyFilter(img, img.Size)
            Return item
        End Function

        Protected Overridable Sub OnPictureEditCurrentImageSizeChanged(ByVal sender As Object, ByVal e As EventArgs)
            If CurrentImage Is Nothing Then Return
            CalcCurrentImageZoomPercent(CurrentImage)
        End Sub

        Protected Overridable Sub OnTbRedEditValueChanged(ByVal sender As Object, ByVal e As EventArgs)
            OnChangeColor()
        End Sub

        Protected Overridable Sub OnTbGreenEditValueChanged(ByVal sender As Object, ByVal e As EventArgs)
            OnChangeColor()
        End Sub

        Protected Overridable Sub OnTbBlueEditValueChanged(ByVal sender As Object, ByVal e As EventArgs)
            OnChangeColor()
        End Sub

        Protected Overridable Sub OnChangeColor()
            If CurrentImage Is Nothing Then Return
            PictureEdit.Image = UniversalFilterInstance.ApplyFilter(CurrentImage, CurrentImage.Size, tbRed.Value, tbGreen.Value, tbBlue.Value)
        End Sub

        Protected Sub OnBrightnessChanging(ByVal sender As Object, ByVal e As XtraEditors.Controls.ChangingEventArgs)
            Dim val As Integer = tbBrightness.Value
            PictureEdit.Image = UniversalFilterInstance.ApplyFilter(CurrentImage, CurrentImage.Size, val, val, val)
        End Sub

        Protected Overridable Sub OnContrastChanged(ByVal sender As Object, ByVal e As EventArgs)
            If CurrentImage Is Nothing Then Return
            Dim val As Integer = tbContrast.Value
            PictureEdit.Image = ContrastFilterInstance.ApplyFilter(CurrentImage, CurrentImage.Size, val)
        End Sub

        Protected Overridable Sub OnTileControlItemClick(ByVal sender As Object, ByVal e As TileItemEventArgs)
            SetImage(e.Item.BackgroundImage)
        End Sub

#Region "Filters"
        Private filterListCore As List(Of FilterBase)

        Protected Overridable ReadOnly Property FilterList As List(Of FilterBase)
            Get
                If filterListCore Is Nothing Then
                    filterListCore = New List(Of FilterBase)() From {New PolaroidFilter(), New BGRFilter(), New GBRFilter(), New SepiaFilter(), New GrayScaleFilter(), New NegativeFilter()}
                End If

                Return filterListCore
            End Get
        End Property

        Private universalFilter As UniversalFilter

        Protected Overridable ReadOnly Property UniversalFilterInstance As UniversalFilter
            Get
                If universalFilter Is Nothing Then universalFilter = New UniversalFilter()
                Return universalFilter
            End Get
        End Property

        Private contrastFilter As ContrastFilter

        Protected Overridable ReadOnly Property ContrastFilterInstance As ContrastFilter
            Get
                If contrastFilter Is Nothing Then contrastFilter = New ContrastFilter()
                Return contrastFilter
            End Get
        End Property
#End Region
    End Class

    '
    Public Module ImageHelper

        Public Function GetDimension(ByVal img As Image) As String
            If img Is Nothing Then Return "0x0"
            Return String.Format("{0}x{1}", img.Width, img.Height)
        End Function

        Public Function GetName(ByVal img As Image) As String
            If img Is Nothing Then Return String.Empty
            Return Path.GetFileName(img.Tag.ToString())
        End Function

        Public Function GetSize(ByVal img As Image) As String
            If img Is Nothing Then Return "0 KB"
            Return(New FileInfo(img.Tag.ToString()).Length / 1024).ToString() & " KB"
        End Function

        Public Function GetRating(ByVal img As Image) As Integer
            If img Is Nothing Then Return 0
            Return TutorialConstants.Random.Next(5)
        End Function

        Private ReadOnly images As Dictionary(Of String, Bitmap) = New Dictionary(Of String, Bitmap)()

        Public Function CreateSenderImage(ByVal sender As String, ByVal size As Size, ByVal ulf As LookAndFeel.UserLookAndFeel) As Bitmap
            Dim img As Bitmap = Nothing
            If images.TryGetValue(sender, img) Then Return img
            Dim imageSize As Size = New Size(size.Width, size.Height)
            Dim parsedText As String() = sender.Split(" "c)
            Dim result As String = String.Empty
            Dim count As Integer = 0
            For i As Integer = 0 To parsedText.Length - 1
                If parsedText(i).Length > 0 Then
                    result += parsedText(i)(0)
                    count += 1
                    If count = 2 Then Exit For
                End If
            Next

            img = GlyphPainter.CreateRoundedStubGlyph(ulf, imageSize, result)
            images.Add(sender, img)
            Return img
        End Function
    End Module

    Public Delegate Sub GalleryItemApplyFilter(ByVal name As String, ByVal image As Image, ByVal filter As FilterBase)

    Public Class FilterBase

        Public Overridable ReadOnly Property Name As String
            Get
                Return String.Empty
            End Get
        End Property

        Public Function ApplyFilter(ByVal image As Image, ByVal size As Size) As Image
            Return ApplyMatrix(image, size, Matrix)
        End Function

        Protected Overridable ReadOnly Property Matrix As ColorMatrix
            Get
                Return New ColorMatrix()
            End Get
        End Property

        Protected Function ApplyMatrix(ByVal image As Image, ByVal size As Size, ByVal matrix As ColorMatrix) As Bitmap
            Dim src As Bitmap = New Bitmap(image, size)
            Dim dest As Bitmap = New Bitmap(src.Width, src.Height, PixelFormat.Format32bppArgb)
            Using graphics As Graphics = Graphics.FromImage(dest)
                Dim bmpAttributes As ImageAttributes = New ImageAttributes()
                bmpAttributes.SetColorMatrix(matrix)
                graphics.DrawImage(src, New Rectangle(0, 0, src.Width, src.Height), 0, 0, src.Width, src.Height, GraphicsUnit.Pixel, bmpAttributes)
            End Using

            src.Dispose()
            Return dest
        End Function
    End Class

    Public Class PolaroidFilter
        Inherits FilterBase

        Public Overrides ReadOnly Property Name As String
            Get
                Return "Polaroid"
            End Get
        End Property

        Protected Overrides ReadOnly Property Matrix As ColorMatrix
            Get
                Return New ColorMatrix(New Single()() {New Single() {1.438F, -0.062F, -0.062F, 0, 0}, New Single() {-0.122F, 1.378F, -0.122F, 0, 0}, New Single() {0.016F, -0.016F, 1.438F, 0, 0}, New Single() {0, 0, 0, 1, 0}, New Single() {0.03F, 0.05F, -0.2F, 0, 1}})
            End Get
        End Property
    End Class

    Public Class GrayScaleFilter
        Inherits FilterBase

        Public Overrides ReadOnly Property Name As String
            Get
                Return "GrayScale"
            End Get
        End Property

        Protected Overrides ReadOnly Property Matrix As ColorMatrix
            Get
                Return New ColorMatrix(New Single()() {New Single() {.3F, .3F, .3F, 0, 0}, New Single() {.59F, .59F, .59F, 0, 0}, New Single() {.11F, .11F, .11F, 0, 0}, New Single() {0, 0, 0, 1, 0}, New Single() {0, 0, 0, 0, 1}})
            End Get
        End Property
    End Class

    Public Class NegativeFilter
        Inherits FilterBase

        Public Overrides ReadOnly Property Name As String
            Get
                Return "Negative"
            End Get
        End Property

        Protected Overrides ReadOnly Property Matrix As ColorMatrix
            Get
                Return New ColorMatrix(New Single()() {New Single() {-1, 0, 0, 0, 0}, New Single() {0, -1, 0, 0, 0}, New Single() {0, 0, -1, 0, 0}, New Single() {0, 0, 0, 1, 0}, New Single() {1, 1, 1, 0, 1}})
            End Get
        End Property
    End Class

    Public Class SepiaFilter
        Inherits FilterBase

        Public Overrides ReadOnly Property Name As String
            Get
                Return "Sepia"
            End Get
        End Property

        Protected Overrides ReadOnly Property Matrix As ColorMatrix
            Get
                Return New ColorMatrix(New Single()() {New Single() {.393F, .349F, .272F, 0, 0}, New Single() {.769F, .686F, .534F, 0, 0}, New Single() {.189F, .168F, .131F, 0, 0}, New Single() {0, 0, 0, 1, 0}, New Single() {0, 0, 0, 0, 1}})
            End Get
        End Property
    End Class

    Public Class BGRFilter
        Inherits FilterBase

        Public Overrides ReadOnly Property Name As String
            Get
                Return "BGR"
            End Get
        End Property

        Protected Overrides ReadOnly Property Matrix As ColorMatrix
            Get
                Return New ColorMatrix(New Single()() {New Single() {0, 0, 1, 0, 0}, New Single() {0, 1, 0, 0, 0}, New Single() {1, 0, 0, 0, 0}, New Single() {0, 0, 0, 1, 0}, New Single() {0, 0, 0, 0, 1}})
            End Get
        End Property
    End Class

    Public Class GBRFilter
        Inherits FilterBase

        Public Overrides ReadOnly Property Name As String
            Get
                Return "GBR"
            End Get
        End Property

        Protected Overrides ReadOnly Property Matrix As ColorMatrix
            Get
                Return New ColorMatrix(New Single()() {New Single() {0, 1, 0, 0, 0}, New Single() {0, 0, 1, 0, 0}, New Single() {1, 0, 0, 0, 0}, New Single() {0, 0, 0, 1, 0}, New Single() {0, 0, 0, 0, 1}})
            End Get
        End Property
    End Class

    Public Class UniversalFilter
        Inherits FilterBase

        Public Sub New()
            b = 0
            g = b
            r = g
        End Sub

        Private r, g, b As Integer

        Public Overloads Function ApplyFilter(ByVal image As Image, ByVal size As Size, ByVal r As Integer, ByVal g As Integer, ByVal b As Integer) As Image
            Me.r = r
            Me.g = g
            Me.b = b
            Return ApplyMatrix(image, size, Matrix)
        End Function

        Public Overrides ReadOnly Property Name As String
            Get
                Return "Universal"
            End Get
        End Property

        Protected Overrides ReadOnly Property Matrix As ColorMatrix
            Get
                Return New ColorMatrix(New Single()() {New Single() {1 + CSng(r / 255.0F), 0, 0, 0, 0}, New Single() {0, 1 + CSng(g / 255.0F), 0, 0, 0}, New Single() {0, 0, 1 + CSng(b / 255.0F), 0, 0}, New Single() {0, 0, 0, 1, 0}, New Single() {0.1F, 0.1F, 0.1F, 0, 1}})
            End Get
        End Property
    End Class

    Public Class ContrastFilter
        Inherits FilterBase

        Public Sub New()
            translate = 0
            scale = translate
        End Sub

        Private scale As Single

        Private translate As Single

        Public Overrides ReadOnly Property Name As String
            Get
                Return "Contrast"
            End Get
        End Property

        Public Overloads Function ApplyFilter(ByVal image As Image, ByVal size As Size, ByVal val As Integer) As Image
            scale = val
            translate =(-.5F * scale + .5F) * 255.0F
            Return ApplyMatrix(image, size, Matrix)
        End Function

        Protected Overrides ReadOnly Property Matrix As ColorMatrix
            Get
                Return New ColorMatrix(New Single()() {New Single() {1 + scale / 100, 0, 0, 0, translate}, New Single() {0, 1 + scale / 100, 0, 0, translate}, New Single() {0, 0, 1 + scale / 100, 0, translate}, New Single() {0, 0, 0, 1, 0}, New Single() {0, 0, 0, 0, 1}})
            End Get
        End Property
    End Class
End Namespace

File: WinForms/CS/NavBarMainDemo/Modules/MultimediaModule.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Drawing.Imaging;
using System.IO;
using System.Windows.Forms;
using DevExpress.DXperience.Demos;
using DevExpress.Utils.Drawing;
using DevExpress.XtraBars.Ribbon;
using DevExpress.XtraEditors;

namespace DevExpress.XtraNavBar.Demos {
    public partial class AccordionControlMultimediaModule : TutorialControl {
        public AccordionControlMultimediaModule() {
            InitializeComponent();
            WorkingFolder = DataPath;
            if(!string.IsNullOrEmpty(WorkingFolder)) {
                InitializeContent();
                SetImage(Images[0]);
            }
        }
        protected virtual void InitializeContent() {
            InitializeTileControl();
            UpdateEditors();
            galleryFilters.SizeChanged += OnGalleryFiltersSizeChanged;
        }
        protected virtual void InitializeTileControl() {
            (tileControl1 as ITileControl).ViewInfo.UseAdvancedTextRendering = false;
            if(TileGroupImages == null)
                return;
            TileControl.Groups.Clear();
            TileControl.Groups.Add(TileGroupImages);
        }
        protected virtual void InitializeImages() {
            if(string.IsNullOrEmpty(WorkingFolder))
                return;
            Images = new List<Image>();
            string[] files = Directory.GetFiles(WorkingFolder, "*.jpg");
            foreach(string file in files) {
                Image img = Bitmap.FromFile(file);
                img.Tag = file;
                Images.Add(img);
            }
            InitializeTileControl();
        }
        protected virtual List<Image> Images { get; set; }
        protected virtual TileControl TileControl {
            get { return tileControl1; }
        }
        protected virtual PictureEdit PictureEdit {
            get { return peCurrentImage; }
        }
        protected virtual GalleryItemGroup FilterGroup {
            get { return galleryFilters.Gallery.Groups[0]; }
        }
        protected virtual Size PreviewSize {
            get { return galleryFilters.Gallery.ImageSize; }
        }
        Image originalImageCore;
        protected virtual Image OriginalImage {
            get { return originalImageCore; }
            set {
                originalImageCore = value;
                peOriginalImage.Image = originalImageCore;
                UpdateImageProperties();
            }
        }
        Image currentImageCore;
        protected virtual Image CurrentImage {
            get { return currentImageCore; }
            set {
                currentImageCore = value;
                if(CurrentImage == null) return;
                PictureEdit.Image = CurrentImage;
                CalcCurrentImageZoomPercent(currentImageCore);
                UpdateEditors();
            }
        }
        protected virtual void CalcCurrentImageZoomPercent(Image img) {
            PictureEdit.Properties.ZoomPercent = Math.Min(((double)PictureEdit.Height / (double)img.Height) * 100, 100);
        }
        string workingFolderCore;
        protected virtual string WorkingFolder {
            get { return workingFolderCore; }
            set {
                if(workingFolderCore == value) return;
                workingFolderCore = value;
                InitializeImages();
            }
        }
        TileGroup group;
        protected virtual TileGroup TileGroupImages {
            get {
                if(group == null)
                    group = CreateTileGroupImages();
                return group;
            }
        }
        protected virtual void SetImage(Image img) {
            CurrentImage = OriginalImage = img;
            RefreshFilters(OriginalImage);
        }
        protected string DataPath {
            get {
                string dataDir = Tutorials.FilePathUtils.FindDirPath("Data");
                string dataPathCore = Path.Combine(dataDir, "AccordionControlData");
                if(Directory.Exists(dataPathCore))
                    return dataPathCore;
                return string.Empty;
            }
        }
        protected virtual void UpdateEditors() {
            bool value = CurrentImage != null;
            tbRed.Enabled = value;
            tbGreen.Enabled = value;
            tbBlue.Enabled = value;
            tbBrightness.Enabled = value;
            tbContrast.Enabled = value;

            tbRed.Value = tbRed.Properties.Minimum;
            tbGreen.Value = tbGreen.Properties.Minimum;
            tbBlue.Value = tbBlue.Properties.Minimum;
            tbBrightness.Value = tbBrightness.Properties.Minimum;
            tbContrast.Value = tbContrast.Properties.Minimum;
        }
        protected virtual TileGroup CreateTileGroupImages() {
            if(Images == null || Images.Count == 0)
                return null;
            TileGroup g = new TileGroup();
            foreach(Image img in Images) {
                TileItem item = new TileItem() {
                    BackgroundImage = img,
                    BackgroundImageScaleMode = TileItemImageScaleMode.Squeeze,
                    BackgroundImageAlignment = TileItemContentAlignment.MiddleCenter
                };
                g.Items.Add(item);
            }
            return g;
        }
        protected virtual void RefreshFilters(Image img) {
            FilterGroup.Items.Clear();
            foreach(FilterBase filter in FilterList) {
                FilterGroup.Items.Add(CreateGalleryItemFilter(img, filter));
            }
        }
        bool updatingAccordionWidth;
        protected virtual void OnGalleryFiltersSizeChanged(object sender, EventArgs e) {
            UpdateAccordionPanelWidth();
        }
        protected virtual void UpdateAccordionPanelWidth() {
            if(updatingAccordionWidth) return;
            if(FilterGroup.Items.Count == 0) return;
            updatingAccordionWidth = true;
            try {
                Size bestSize = galleryFilters.GetPreferredSize(Size.Empty);
                int delta = bestSize.Width - galleryFilters.Width;
                if(delta == 0) return;
                panelControl7.Width += delta;
            }
            finally {
                updatingAccordionWidth = false;
            }
        }
        protected virtual void UpdateImageProperties() {
            ratingImage.Rating = ImageHelper.GetRating(OriginalImage);
            lblSize.Text = ImageHelper.GetSize(OriginalImage);
            lblName.Text = ImageHelper.GetName(OriginalImage);
            lblDimension.Text = ImageHelper.GetDimension(OriginalImage);
        }
        protected virtual GalleryItem CreateGalleryItemFilter(Image img, FilterBase filter) {
            GalleryItem item = new GalleryItem() {
                Image = filter.ApplyFilter(img, PreviewSize),
                Caption = filter.Name,
            };
            item.ItemClick += delegate { PictureEdit.Image = filter.ApplyFilter(img, img.Size); };
            return item;
        }
        protected virtual void OnPictureEditCurrentImageSizeChanged(object sender, EventArgs e) {
            if(CurrentImage == null)
                return;
            CalcCurrentImageZoomPercent(CurrentImage);
        }
        protected virtual void OnTbRedEditValueChanged(object sender, EventArgs e) { OnChangeColor(); }
        protected virtual void OnTbGreenEditValueChanged(object sender, EventArgs e) { OnChangeColor(); }
        protected virtual void OnTbBlueEditValueChanged(object sender, EventArgs e) { OnChangeColor(); }
        protected virtual void OnChangeColor() {
            if(CurrentImage == null)
                return;
            PictureEdit.Image = UniversalFilterInstance.ApplyFilter(CurrentImage, CurrentImage.Size, tbRed.Value, tbGreen.Value, tbBlue.Value);
        }
        protected void OnBrightnessChanging(object sender, DevExpress.XtraEditors.Controls.ChangingEventArgs e) {
            int val = tbBrightness.Value;
            PictureEdit.Image = UniversalFilterInstance.ApplyFilter(CurrentImage, CurrentImage.Size, val, val, val);
        }
        protected virtual void OnContrastChanged(object sender, EventArgs e) {
            if(CurrentImage == null)
                return;
            int val = tbContrast.Value;
            PictureEdit.Image = ContrastFilterInstance.ApplyFilter(CurrentImage, CurrentImage.Size, val);
        }
        protected virtual void OnTileControlItemClick(object sender, TileItemEventArgs e) {
            SetImage(e.Item.BackgroundImage);
        }
        #region Filters
        List<FilterBase> filterListCore;
        protected virtual List<FilterBase> FilterList {
            get {
                if(filterListCore == null) {
                    filterListCore = new List<FilterBase>(){
                        new PolaroidFilter(),
                        new BGRFilter(),
                        new GBRFilter(),
                        new SepiaFilter(),
                        new GrayScaleFilter(),
                        new NegativeFilter(),
                    };
                }
                return filterListCore;
            }
        }
        UniversalFilter universalFilter;
        protected virtual UniversalFilter UniversalFilterInstance {
            get {
                if(universalFilter == null)
                    universalFilter = new UniversalFilter();
                return universalFilter;
            }
        }
        ContrastFilter contrastFilter;
        protected virtual ContrastFilter ContrastFilterInstance {
            get {
                if(contrastFilter == null)
                    contrastFilter = new ContrastFilter();
                return contrastFilter;
            }
        }
        #endregion
    }
    //
    public static class ImageHelper {
        public static string GetDimension(Image img) {
            if(img == null) return "0x0";
            return string.Format("{0}x{1}", img.Width, img.Height);
        }
        public static string GetName(Image img) {
            if(img == null) return string.Empty;
            return Path.GetFileName(img.Tag.ToString());
        }
        public static string GetSize(Image img) {
            if(img == null) return "0 KB";
            return (new FileInfo(img.Tag.ToString()).Length / 1024).ToString() + " KB";
        }
        public static int GetRating(Image img) {
            if(img == null) return 0;
            return TutorialConstants.Random.Next(5);
        }
        readonly static Dictionary<string, Bitmap> images = new Dictionary<string, Bitmap>();
        public static Bitmap CreateSenderImage(string sender, Size size, DevExpress.LookAndFeel.UserLookAndFeel ulf) {
            Bitmap img = null;
            if(images.TryGetValue(sender, out img))
                return img;
            Size imageSize = new Size(size.Width, size.Height);
            string[] parsedText = sender.Split(' ');
            string result = string.Empty;
            int count = 0;
            for(int i = 0; i < parsedText.Length; i++) {
                if(parsedText[i].Length > 0) {
                    result += parsedText[i][0];
                    count++;
                    if(count == 2) break;
                }
            }
            img = GlyphPainter.CreateRoundedStubGlyph(ulf, imageSize, result);
            images.Add(sender, img);
            return img;
        }
    }
    public delegate void GalleryItemApplyFilter(string name, Image image, FilterBase filter);
    public class FilterBase {
        public virtual string Name { get { return string.Empty; } }
        public Image ApplyFilter(Image image, Size size) {
            return ApplyMatrix(image, size, Matrix);
        }
        protected virtual ColorMatrix Matrix { get { return new ColorMatrix(); } }
        protected Bitmap ApplyMatrix(Image image, Size size, ColorMatrix matrix) {
            Bitmap src = new Bitmap(image, size);
            Bitmap dest = new Bitmap(src.Width, src.Height, PixelFormat.Format32bppArgb);
            using(Graphics graphics = Graphics.FromImage(dest)) {
                ImageAttributes bmpAttributes = new ImageAttributes();
                bmpAttributes.SetColorMatrix(matrix);
                graphics.DrawImage(src, new Rectangle(0, 0, src.Width, src.Height), 0, 0, src.Width, src.Height, GraphicsUnit.Pixel, bmpAttributes);
            }
            src.Dispose();
            return dest;
        }
    }
    public class PolaroidFilter : FilterBase {
        public override string Name { get { return "Polaroid"; } }
        protected override ColorMatrix Matrix {
            get {
                return new ColorMatrix(new float[][] {
                    new float[] {1.438f, -0.062f, -0.062f, 0, 0},
                    new float[] {-0.122f, 1.378f, -0.122f, 0, 0},
                    new float[] {0.016f, -0.016f, 1.438f, 0, 0},
                    new float[] {0, 0, 0, 1, 0},
                    new float[] {0.03f, 0.05f, -0.2f, 0, 1}
                });
            }
        }
    }
    public class GrayScaleFilter : FilterBase {
        public override string Name { get { return "GrayScale"; } }
        protected override ColorMatrix Matrix {
            get {
                return new ColorMatrix(new float[][] {
                    new float[] {.3f, .3f, .3f, 0, 0},
                    new float[] {.59f, .59f, .59f, 0, 0},
                    new float[] {.11f, .11f, .11f, 0, 0},
                    new float[] {0, 0, 0, 1, 0},
                    new float[] {0, 0, 0, 0, 1}
                });
            }
        }
    }
    public class NegativeFilter : FilterBase {
        public override string Name { get { return "Negative"; } }
        protected override ColorMatrix Matrix {
            get {
                return new ColorMatrix(new float[][] {
                    new float[] {-1, 0, 0, 0, 0},
                    new float[] {0, -1, 0, 0, 0},
                    new float[] {0, 0, -1, 0, 0},
                    new float[] {0, 0, 0, 1, 0},
                    new float[] {1, 1, 1, 0, 1}
                });
            }
        }
    }
    public class SepiaFilter : FilterBase {
        public override string Name { get { return "Sepia"; } }
        protected override ColorMatrix Matrix {
            get {
                return new ColorMatrix(new float[][] {
                    new float[] {.393f, .349f, .272f, 0, 0},
                    new float[] {.769f, .686f, .534f, 0, 0},
                    new float[] {.189f, .168f, .131f, 0, 0},
                    new float[] {0, 0, 0, 1, 0},
                    new float[] {0, 0, 0, 0, 1}
                });
            }
        }
    }
    public class BGRFilter : FilterBase {
        public override string Name { get { return "BGR"; } }
        protected override ColorMatrix Matrix {
            get {
                return new ColorMatrix(new float[][] {
                    new float[] {0, 0, 1, 0, 0},
                    new float[] {0, 1, 0, 0, 0},
                    new float[] {1, 0, 0, 0, 0},
                    new float[] {0, 0, 0, 1, 0},
                    new float[] {0, 0, 0, 0, 1}
                });
            }
        }
    }
    public class GBRFilter : FilterBase {
        public override string Name { get { return "GBR"; } }
        protected override ColorMatrix Matrix {
            get {
                return new ColorMatrix(new float[][] {
                    new float[] {0, 1, 0, 0, 0},
                    new float[] {0, 0, 1, 0, 0},
                    new float[] {1, 0, 0, 0, 0},
                    new float[] {0, 0, 0, 1, 0},
                    new float[] {0, 0, 0, 0, 1}
                });
            }
        }
    }
    public class UniversalFilter : FilterBase {
        public UniversalFilter() {
            r = g = b = 0;
        }
        int r, g, b;
        public Image ApplyFilter(Image image, Size size, int r, int g, int b) {
            this.r = r;
            this.g = g;
            this.b = b;
            return ApplyMatrix(image, size, Matrix);
        }
        public override string Name { get { return "Universal"; } }
        protected override ColorMatrix Matrix {
            get {
                return new ColorMatrix(new float[][] {
                    new float[] {1 + (float)(r / 255.0f), 0, 0, 0, 0},
                    new float[] {0, 1 + (float)(g/ 255.0f), 0, 0, 0},
                    new float[] {0, 0, 1 + (float)(b/ 255.0f), 0, 0},
                    new float[] {0, 0, 0, 1, 0},
                    new float[] {0.1f, 0.1f, 0.1f, 0, 1}
                });
            }
        }
    }
    public class ContrastFilter : FilterBase {
        public ContrastFilter() {
            scale = translate = 0;
        }
        float scale;
        float translate;
        public override string Name { get { return "Contrast"; } }
        public Image ApplyFilter(Image image, Size size, int val) {
            scale = val;
            translate = (-.5f * scale + .5f) * 255.0f;
            return ApplyMatrix(image, size, Matrix);
        }
        protected override ColorMatrix Matrix {
            get {
                return new ColorMatrix(new float[][] {
                    new float[] {1 + scale / 100, 0, 0, 0, translate},
                    new float[] {0, 1 + scale / 100, 0, 0, translate},
                    new float[] {0, 0, 1 + scale / 100, 0, translate},
                    new float[] {0, 0, 0, 1, 0},
                    new float[] {0, 0, 0, 0, 1}
                });
            }
        }
    }
}
